98_Drivers/Timers.md

```diff
@@ -0,0 +1,283 @@
+
+We're mainly interested in the `address` field which gives us the physical address of the HPET registers. The other fields are explained in the HPET specification but are not needed for our purposes right now.
+
+As with any MMIO you will need to map this physical address into the virtual address space so we can access the registers with paging enabled.
+
+### Theory Of Operation
+
+The HPET consists of a single main counter (that counts up) with some global configuration, and a number of comparators that can trigger interrupts when certain conditions are met in relation to the main counter. The HPET will always have at least 3 comparators, but may have up to 32.
+
+The HPET is similar to the PIT in that we are told the frequency of it's clock. Unlike the PIT, the HPET spec does not give us the frequency directly, we have to read it from the HPET registers. 
+
+Each register is accessed by adding an offset to the base address we obtained before. The main registers we're interested in are:
+
+- General capabilities: offset 0x0.
+- General configuration: offset 0x10.
+- Main counter value: 0xF0.
+
+We can read the main counter at any time, which is measured in in timer ticks. We can convert these ticks into realtime by multiplying them with the timer period in the general capabilities register. Bits 63:32 of the general capabilities register contain the number of femtoseconds for each tick. A nanosecond is 1000 femtoseconds, and 1 second is 1'000'000'000 femtoseconds.
+
+We can also write to the main counter, usually you would write a 0 here when initializing the HPET in order to be able to determine uptime, but this is not really necessary.
+
+The general capabilities register contains some other useful information, like whether the main counter is 32 or 64 bits wide, and the number of comparators (which are used to generate interrupts). These are all detailed in the public spec, and you can explore these at your leisure.
+
+In order for the main counter to actually begin counting, we need to enable it. This is done by setting bit 0 of the general configuration register. Once this bit is set, the main counter will increment by one every time it's internal clock ticks. The period of this clock is what's specified in the general capabilities register (bits 63:32).
+
+### Comparators
+
+The main counter is only suitable for polling the time, but it cannot generate interrupts. For that we have to use one of the comparators. The HPET will always have at least three comparators, but may have up to 32. In reality most vendors use the stock intel chip which comes with 3 comparators, although some older and more exotic hardware may provide more.
```

Comment:
The HPET will always have at least three comparators but may have up to 32, 

this is the same sentence writtien in the previous paragraphs, is fine repeating if necessary, do you think could be better if that can be reworded, or is not necessary?